File: python/search_in_rotated_sorted_array.py
class Solution:
    # @param A, a list of integers
    # @param target, an integer to be searched
    # @return an integer
    def search(self, A, target):
        if not A:
            return -1
        lo, hi = 0, len(A)-1
        while lo <= hi:
            mid = (lo+hi) // 2
            midval = A[mid]
            if midval == target:
                return mid
            elif A[lo] <= midval:
                # the bottom half is sorted
                if A[lo] <= target < midval:
                    hi = mid - 1
                else:
                    lo = mid + 1
            else:
                if midval < target <= A[hi]:
                    lo = mid + 1
                else:
                    hi = mid - 1
        return -1

File: java/SearchinRotatedSortedArray.java
/**
 * Created by Robbert on 1/26/15.
 * Search in Rotated Sorted Array
 * https://oj.leetcode.com/problems/search-in-rotated-sorted-array/
 */
public class SearchinRotatedSortedArray {
    public int search(int[] A, int target) {
        if (A == null || A.length == 0) {
            return -1;
        }

        int lo = 0;
        int hi = A.length - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (A[mid] == target) {
                return mid;
            }

            if (A[lo] <= A[mid]) {
                if (A[lo] <= target && target < A[mid]) {
                    hi = mid - 1;
                } else {
                    lo = mid + 1;
                }
            } else {
                if (A[mid] < target && target <= A[hi]) {
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }

        }
        return -1;
    }
}
